### ISSUE TITLE
No way to set option eagerEsModules in swc relay plugin causing error in recoil and relay app

### ISSUE THREAD

**Issue by johnsonjo4531** [swc-project/plugins#107]
tl;dr: swc-relay doesn't appear to make the option available for eagerEsModules for the relay-compiler. This seems to be causing the issue in my recoil and relay app:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'operationKind')
```

I'm using recoil and relay together and I'm having this issue as raised [here](https://github.com/facebookexperimental/Recoil/issues/1998).

I've create [a repo with reproduction of the issue](https://github.com/johnsonjo4531/github-relay-app-swc-error) in swc, as well as in babel, and with a fix in babel that isn't possible in swc yet (the fix comes from the issue above raised in recoil).

I've created three seperate branches in this repository other than the `main` branch (the main branch is however uneccessary to look at as it's just the getting started guide for relay.) 

The three branches to lookat are called: `webpack-babel-relay`, `webpack-babel-relay-failing`, and `webpack-swc-relay`.

> IMPORTANT NOTE: in order to setup the three branches of the repo to work properly you have to setup your environment variables by adding a github accesstoken. Follow [the steps here in the relay getting started guide to do so](https://relay.dev/docs/getting-started/step-by-step-guide/#21-github-graphql-authentication). 

`webpack-babel-relay` is the only branch working of the three mentioned (with recoil atleast). It properly functions and produces the proper result from the following query:

```tsx
// ...
const RepositoryNameQuery = graphql`
  query AppRepositoryNameQuery {
    repository(owner: "facebook", name: "relay") {
      name
    }
  }
`;

const repositoryNameSelector = graphQLSelector<{}, AppRepositoryNameQuery$data>(
  {
    key: "RepoName",
    query: RepositoryNameQuery,
    environment: myEnvironmentKey,
    variables: {},
    mapResponse: (data) => data,
  }
);

// React component hook 
function App() {
  const data = useRecoilValue(repositoryNameSelector); // <-- this fails in the other two repos.
  // ...
```

The error that is thrown by the other two branches which are not working is:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'operationKind')
```

This is an internal Recoil error, so my code doesn't have much control over it.

`webpack-babel-relay-failing` fails the above code at the line mentioned and the only difference between it and `webpack-babel-relay` is that `webpack-babel-relay` has the eagerEsModules field set to `true` in the relay config in the package.json.

`webpack-swc-relay` fails the above code as well despite having the eagerEsModules field set to true in the relay config in the package.json. I have tried alternatives like setting swc's relay plugin configuration but nothing has seemed to work.

So, I'm thinking that if I could just change the eagerEsModules config in the relay compiler via the swc-relay plugin it would work fine.

P.S. here is [a set of all the other options that can be set by the relay compiler](https://github.com/facebook/relay/tree/main/packages/relay-compiler). I only post them here since they were a little hard to find.

**Comment by johnsonjo4531** [comment]
By the way a bit more info:

I've noticed that another alternative to the above eagerEsModules config fix is to change the above code to:

```ts
const RepositoryNameQuery = graphql`
  query AppRepositoryNameQuery {
    repository(owner: "facebook", name: "relay") {
      name
    }
  }
`;

const repositoryNameSelector = graphQLSelector<{}, AppRepositoryNameQuery$data>(
  {
    key: "RepoName",
    query: RepositoryNameQuery.default, // <--- this line changed to add `.default` after the relay query
    environment: myEnvironmentKey,
    variables: {},
    mapResponse: (data) => data,
  }
);

// React component hook 
function App() {
  const data = useRecoilValue(repositoryNameSelector); // <-- this fails in the other two repos.
  // ...
```

The problem I have with doing the above is that although it works my editor complains about me passing in the wrong type to recoil (I'm using typescript and it's erroring only after the line is changed).

**Comment by EloB** [comment]
Is there any fix to this except using `query.default`?

When I console log the "query" I get:
```
Object [Module] { default: [Getter] }
```

**Comment by johnsonjo4531** [comment]
> Is there any fix to this except using `query.default`?

@EloB Kind of... just not using swc as far as I know, but using babel. I think they would have to add a new swc-relay plugin option for it to work using swc which is why I raised the issue.

The first branch that I said is working is using webpack, babel, and relay. That is the way you could get it to work you have to pass `eagerEsModules: true` to the relay config as I tried to specify here:

> webpack-babel-relay-failing fails the above code at the line mentioned and the only difference between it and webpack-babel-relay is that webpack-babel-relay has the eagerEsModules field set to true in the relay config in the package.json.

**Comment by EloB** [comment]
@johnsonjo4531 Its broken only for Typescript right? I sat all day yesterday trying to find a solution to keep SWC by creating a loader or plugin for webpack to try to auto resolve that default.

Seems like the typescript code is not being transpiled after been into the SWC plugin.

**Comment by johnsonjo4531** [comment]
> How can we solve this? I've tried to make a workaround in Webpack until this is properly fixed but it's not complete.

@EloB I could see any of the following possible solutions until this repo pushes a fix. Most of them involve using type casting I think the reusable function may be the way to go:

```tsx
import { graphql, GraphQLTaggedNode } from "react-relay";

const pagesIndexQuery = graphql`
  query pagesIndexQuery @preloadable {
    TweetsMeta {
      count
    }
  }
`;

console.log(
  "test1 (doesn't work the wrong query is returned)",
  pagesIndexQuery
);
console.log(
  "test2 (does work at runtime but the wrong type is returned and thus fails compile time checks.)",
  pagesIndexQuery.default
);
// Solution 1: cast to unknown and then the desired type.
console.log(
  "test3 (does work at runtime but has a lot of type wrangling inline and isn't super reusable.)",
  ((pagesIndexQuery as unknown) as { default: GraphQLTaggedNode }).default
); // I don't want todo this.
// Solution 2: create a reusable function that does the above for you.
/**
 * GraphQLTaggedNode's act bizzare in swc, so this function provides
 * a way to turn them into the actual node itself and keep the type
 * safety. */
function getTaggedNode(query: GraphQLTaggedNode) {
  if ("default" in query) {
    return ((query as unknown) as { default: GraphQLTaggedNode }).default;
  } else {
    return query;
  }
}

// Then you can use the above function like so:
console.log("test4 (does work at runtime)", getTaggedNode(pagesIndexQuery));
// You can even repeatedly use the function and it should work too:
console.log(
  "test4 (still works at runtime)",
  getTaggedNode(getTaggedNode(pagesIndexQuery))
);

export default function IndexPage() {
  return <div>Typescript</div>;
}
```

P.S. for some reason I could not fork your sandbox, so technically I couldn't even really test the above codes for any sort of errors I may have made, so just keep that in mind if there is some sort of error you can feel free to ping me again. :)

**Comment by kdy1** [comment]
Fixed by https://github.com/swc-project/plugins/pull/122
